### DISCUSSION TITLE
New Contributor Setup

### DISCUSSION THREAD

**Discussion by willhansen** [GothenburgBitFactory/taskwarrior#3536] [General]
I've been trying to get the develop branch just compiling and passing tests to get started, and I'm curious how far off I am from the "standard" route.  

I had to tweak a few things in the `docker/task.dockerfile` file. ([diff](https://github.com/GothenburgBitFactory/taskwarrior/commit/d9d046c563cb2b13f8e0f735919cb79b139310b0))

Commands:
``` bash
git clone -b docker-workflow git@github.com:willhansen/taskwarrior.git && \
cd taskwarrior && \
./docker/build_and_run_in_docker.bash
```

In the container:
```bash
./run_tests.bash
```

Passes all tests.

**Comment by djmitche** [comment]
That Dockerfile is used in CI to build a docker image that is pushed to ghcr.io. I think that's why it doesn't care much for caching, and uses the tag `foo`.

That said, I don't actually see a run of that action -- but that's another issue.

Your changes make sense as a way to run the tests from `develop`, but I"m not sure it's the best way to actually do development, just because it will do a clean build on every run.

The "standard" route is documented [here](https://github.com/GothenburgBitFactory/taskwarrior/blob/develop/doc/devel/contrib/development.md), if that helps. It does require a fairly new CMake, a C++ compiler, and a Rust toolchain.

**Reply by willhansen** [reply]
I've got bind-mount-based caching working for the build directory and cargo registry, but there's still some other test build stuff I'm not sure where it's coming from:

![image](https://github.com/GothenburgBitFactory/taskwarrior/assets/2918280/b75f0274-be18-402c-bf68-9abd10db2804)

**Reply by djmitche** [reply]
Hmmm, what backend are you using? That doesn't look too unusual to me

**Reply by djmitche** [reply]
Oh, that's from the Make backend -- https://gitlab.kitware.com/cmake/cmake/-/issues/23689

**Reply by willhansen** [reply]
Don't have to worry about it then.

Managed to polish up the script a little.  Others may find it useful.

[Branch](https://github.com/willhansen/taskwarrior/tree/docker-workflow)

Docs:
```
Usage:

    ./local-docker/build_and_run_in_docker.bash [-h|--help] [-t|--test [STRING]]

Options:
  -h, --help            Show this help text
  -t, --test [STRING]   Run tests after building
                        If STRING is present, only run tests containing STRING in their name
  -i, --interactive     Leave user in bash shell in container afterwards
```

**Reply by djmitche** [reply]
I guess the motivation to use Docker is to not install a C++ compiler, Rust toolchain, etc. in the base system?

@felixschurk what do you think? Would this make sense added to the repo?

**Reply by felixschurk** [reply]
I am wondering what is the difference to the already existing [`.devcontainer`](https://github.com/GothenburgBitFactory/taskwarrior/tree/2bd609afe38516dc7e4112fd6e4e6f62c7103098/.devcontainer) setup?

At least to my understanding the "result" is similar, one has a docker container in which one can do the development - without the need to install CMake, Rust etc. locally.

**Reply by felixschurk** [reply]
> That Dockerfile is used in CI to build a docker image that is pushed to ghcr.io. I think that's why it doesn't care much for caching, and uses the tag `foo`.
> 
> That said, I don't actually see a run of that action -- but that's another issue.
> 
> Your changes make sense as a way to run the tests from `develop`, but I"m not sure it's the best way to actually do development, just because it will do a clean build on every run.
> 
> The "standard" route is documented [here](https://github.com/GothenburgBitFactory/taskwarrior/blob/develop/doc/devel/contrib/development.md), if that helps. It does require a fairly new CMake, a C++ compiler, and a Rust toolchain.

https://github.com/GothenburgBitFactory/taskwarrior/actions/workflows/docker-image.yaml

I think these are the corresponding action runs - however all are basically failed.

**Reply by willhansen** [reply]
> I am wondering what is the difference to the already existing [`.devcontainer`](https://github.com/GothenburgBitFactory/taskwarrior/tree/2bd609afe38516dc7e4112fd6e4e6f62c7103098/.devcontainer) setup?
> 
> At least to my understanding the "result" is similar, one has a docker container in which one can do the development - without the need to install CMake, Rust etc. locally.

Even less than that, really.  I just wanted an easy way to effectively run the CI locally with a one-liner.  

I looked into the `.devcontainer` stuff until I realized it seemed pretty Visual Studio centric and lost interest.

**Reply by felixschurk** [reply]
Ah okay. Now I get the idea.
Devcontainer can also be used with Clion, but the original idea comes from Visual Studio and Visual Studio Code. I do not know which other IDE's also support it.

I think it does not make a lot of sense to add it to the repo, as it would add another 'similar' way of developing and one would also need to be taken care of.
I am also not sure how it would behave with pushing out of the container, I assume one would need to setup new git credentials every time the container is freshly installed?

What we could do is using the existing push to docker action, and maybe provide this docker image to be pulled down?
Such that people could pull latest develop from a docker image?

**Reply by djmitche** [reply]
Over the last few months we've removed a number of off-the-beaten-path things that had fallen out of date, so I agree with not adding a new one. Having a single "well-lit path" for development seems a better choice for a project with this level of activity.

That docker-push idea sounds pretty cool, though!

**Reply by willhansen** [reply]
Big thumbs up for a single well-lit path.  Keeping a bunch of overlapping build systems around would either need way too much maintenance or they'd start to fall by the wayside.
